### ISSUE TITLE
[PHP] Use tree-sitter-php to parse PHP code instead of pfff

### ISSUE THREAD

**Issue by aryx** [semgrep/semgrep#3941]
Pad already done the job of integrating tree-sitter-php in ocaml-tree-sitter-semgrep. We just need now
to finish Parse_php_tree_sitter.ml in semgrep-core/src/parsing/tree-sitter/ and implement all those
todos.

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-412/[php]-use-tree-sitter-php-to-parse-php-code-instead-of-pfff.

**Comment by cfroystad** [comment]
Great! I hope to find some time to complete tree-sitter-php in the not too distant future for our mutual benefit.

Note that there are some AST inconsistencies in the master branch I hope to resolve before the next release and that the [First-class callable syntax](https://wiki.php.net/rfc/first_class_callable_syntax) of PHP 8.1 might (depending on how you extend the grammar) interfere with Semgrep.

**Comment by aryx** [comment]
:wave:

**Comment by Sjord** [comment]
[This comment](https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/src/parsing/tree_sitter/Parse_php_tree_sitter.ml#L28-L32) suggests to map tree-sitter CST to PHP AST, which can then be converted to generic AST. Does this make sense? Wouldn't it be better to map to generic AST immediately?

Currently php_to_generic uses both AST and CST types. E.g. [BoolTy is a CST type](https://github.com/returntocorp/semgrep/blob/develop/semgrep-core/src/parsing/pfff/php_to_generic.ml#L86). But this type is tied to pfff CST, and not to tree-sitter CST. Do we need to map from tree-sitter CST to pfff CST, or should BoolTy be moved to PHP AST, or should we just go directly to generic AST?

**Comment by aryx** [comment]
I would like to remove cst_php.ml from pfff, and only get an ast_php.ml there.
Then, when there is an existing ast_xxx.ml (like the one in pfff), I prefer to translate the tree-sitter CST to this ast_xxx.ml first, and then
separately have a xxx_to_generic.ml.
This is debatable. I usually prefer to separate concerns, but there's a case to be make sometimes to go directly
to the generic AST.
Recently we sometimes went directly to the generic AST (as for Swift, or C#). But recently some people
preferred first to define a proper AST for the language (e.g., AST_bash.ml, AST_dockerfile.ml) and split the process in 2.

**Comment by Sjord** [comment]
I [have started on](https://github.com/returntocorp/semgrep/pull/5264) converting the tree-sitter CST to a [PHP-specific AST](https://github.com/returntocorp/pfff/blob/develop/lang_php/analyze/ast_php.ml). This is going reasonably well, but there is another part I could use some help with. The tree-sitter-php parser in semgrep doesn't seem capable of parsing PHP all that well:

If I have a test.php that contains:
```
<?php
echo "hello world";
```

And I dump the tree-sitter CST through semgrep-core:
```
$ semgrep-core -dump_tree_sitter_cst -l php test.php
|   | "<?php"
|   |   | Exp_stmt
|   |   |   | Un_exp
|   |   |   |   Prim_exp
|   |   |   |   | Choice_int
|   |   |   |   |   Str_
|   |   |   |   |   | Str
|   |   |   |   |   | "\"hello world\""
|   |   |   | SEMI
|   |   |   | ";"
File test.php, line 2, characters 0-4:
<?php
echo "hello world";

parent: program, left sibling: php_tag, first child: name
node type: ERROR
children: [
  name
]
Unrecognized construct
```

It fails to parse the `echo`. Does the tree-sitter-php reference need to be updated? It seems to me that all versions of tree-sitter-php would support `echo` statements. @aryx, do you have any idea?

**Comment by aryx** [comment]
yes maybe we're using an old tree-sitter-php.

**Comment by aryx** [comment]
did you try to call directly tree-sitter-php on those test files? what does tree-sitter dump returns?

**Comment by aryx** [comment]
Apparently we're currently using tree-sitter-php from Aug 1 2021 in ocaml-tree-sitter-semgrep (actually that was a commit from you Sjoerd in tree-sitter-php itself), so maybe this was fixed in later versions.

**Comment by aryx** [comment]
hmm I think the problem comes from ocaml-tree-sitter. This file seems to parse file with tree-sitter-php, just that
via ocaml-tree-sitter we don't get the right CST.

**Comment by aryx** [comment]
The problem seems to come from the use of this keyword() function in grammar.js. If you remove it around 'echo' in
the echo_statement rule, then ocaml-tree-sitter can parse your example fine.

**Comment by aryx** [comment]
@Sjord I've created an issue to fix the root problem here: https://github.com/returntocorp/ocaml-tree-sitter-core/issues/34

**Comment by aryx** [comment]
ok martin fixed the issue, we are now at 99.97% for the parsing-stats in ocaml-tree-sitter-semgrep for PHP!
But we are at 92% in semgrep-core, because of some remaining TODOs I guess. I'll try to fix that next week.

**Comment by aryx** [comment]
ok after the few fixes by @Sjord we are at 98.97% on semgrep-core side for PHP!
